babe: report equivocations

andresilva opened the pull request:
This PR adds support for submitting equivocation reports for BABE. An equivocation in BABE is defined as an authority authoring two different blocks on the same slot. When importing blocks we maintain an index `SlotNumber -> Vec<Header>` and check against this index to figure out if any other block has been authored at the given slot and if so we create an equivocation proof using those headers.

The BABE pallet now has methods `report_equivocation` and `report_equivocation_unsigned` which given an equivocation proof and a key ownership proof will report the offence and reward the reporter. Additionally BABE's runtime API is extended with `submit_report_equivocation_unsigned_extrinsic` and `generate_key_ownership_proof` methods which are meant to be used by the client (in an offchain context) for submitting the equivocation reports.

Unsigned equivocation reports are meant to be sent only by authorities, a `ValidateUnsigned` instance verifies that only local reports can be included in the pool and they are not broadcast to the network.

Currently the client will only send unsigned transactions for reporting equivocations as part of block verification. Extending support to let it use any account to report equivocations through a signed transaction will be implemented later.

cc @jacogr This PR exposes some new types.

polkadot companion: https://github.com/paritytech/polkadot/pull/1330

TODO:
- [x] Figure out weights
- [x] Polkadot companion
- [x] Fix tx submission on future context (https://github.com/paritytech/substrate/pull/6325)
- [x] Docs
- [x] Tests

rphmeier commented on client/consensus/babe/src/lib.rs:
```diff
@@ -757,13 +759,95 @@ impl<Block, Client> BabeVerifier<Block, Client>
+		// state to be able to generate the proof). this might fail if the
+		// equivocation happens on the first block of the session, in which case
+		// its parent would be on the previous session. if generation on the
+		// parent header fails we try with best block as well.
```
Note #6220 is almost in ;)

andresilva replied:
Seems like it's in now but I'd prefer to keep the current approach in this PR and having the change to generate session proofs with OCW be done in a follow-up PR.

rphmeier replied:
Sounds good.

rphmeier commented on bin/node/runtime/src/lib.rs:
```diff
@@ -986,6 +1001,28 @@ impl_runtime_apis! {
 		}
+
+		fn generate_key_ownership_proof(
+			authority_id: sp_consensus_babe::AuthorityId,
```
As with grandpa, we probably want to include the `SessionIndex` or parent hash here

rphmeier replied:
or slot number would be the most domain-specific sensible example

rphmeier commented on frame/babe/src/equivocation.rs:
```diff
@@ -0,0 +1,271 @@
+/// When a validator released two or more blocks at the same slot.
+pub struct BabeEquivocationOffence<FullIdentification> {
+	/// A babe slot number in which this incident happened.
+	pub slot: u64,
```
Should this be the `SlotNumber` type?

rphmeier commented on frame/babe/src/lib.rs:
```diff
@@ -561,6 +597,69 @@ impl<T: Trait> Module<T> {
+
+		let epoch_index = (slot_number.saturating_sub(GenesisSlot::get()) /
+			T::EpochDuration::get())
+		.saturated_into::<u32>();
```
indentation seems weird here

rphmeier commented on primitives/core/src/lib.rs:
```diff
@@ -331,6 +331,11 @@ pub fn to_substrate_wasm_fn_return_value(value: &impl Encode) -> u64 {
+/// The void type - it cannot exist.
+// Oh rust, you crack me up...
+#[derive(Clone, Decode, Encode, Eq, PartialEq, RuntimeDebug)]
+pub enum Void {}
```
I see this is used in this PR, can you explain a bit more about why this is introduced vs. using `()` as is often done?

I see `Void` could be used as a more explicit replacement for `()`.

andresilva replied:
`KeyOwnerProofSystem` is implemented for `()` with the `KeyOwnerProof` type set to `Void` since no key owner proof can be constructed. I moved `Void` from `frame/support` to `primitives/core` so that it can be used in `primitives/session` for implementing the utility traits `GetValidatorCount` and `GetSessionNumber` on it.

rphmeier commented on primitives/consensus/babe/src/lib.rs:
```diff
@@ -237,6 +244,87 @@ pub struct BabeEpochConfiguration {
+/// Verifies the equivocation proof by making sure that: both headers have
+/// different hashes, are targetting the same slot, and have valid signatures by
+/// the same authority.
+pub fn check_equivocation_proof<H>(proof: EquivocationProof<H>) -> Option<()>
```
API nit: a bit weird to use `Option` for this one. I'd expect either `Result<(), ()>` or `bool`

burdges commented:
How big are the reports gossiped here?

rphmeier commented:
@burdges I don't believe the reports are gossipped. We are expecting the block authors themselves to include the reports.

andresilva commented:
The size of the `EquivocationProof` should be around 600 bytes for VRF slot headers that don't include any extra digest (required digests for BABE are the seal and preruntime digest with slot claim proof).

burdges commented:
Should gossip messages that depend upon an `EquivocationProof` carry it with them or can they reference the `EquivocationProof` somehow?

andresilva commented:
Currently there are no gossip messages referencing the equivocation proofs. The offence reporting is done by the block authors by including a transaction in the block that does the reporting and verifies the proof on-chain. Currently this transaction isn't gossiped, the block author is incentivized to include it in order to get the reward.

andresilva commented:
I added some benchmarks and added the weight for the report equivocation extrinsic based on those. I have only run them locally for now (will update the numbers tomorrow after running on baseline hardware). The PR should be ready for review now.

cecton commented:
( @andresilva the fix for the CI has been merged, you can update your branch )

andresilva commented:
Ran the benchmarks on the reference hardware and made some small changes to the weights. Should be ready for final review.

octol commented on client/consensus/babe/src/lib.rs:
```diff
@@ -757,13 +759,95 @@ impl<Block, Client> BabeVerifier<Block, Client>
+
+		// don't report any equivocations during initial sync
+		// as they are most likely stale.
+		if *origin == BlockOrigin::NetworkInitialSync {
+			return Ok(());
+		}
```
Tiny nitpick: 
A simple early return like this could go first in function? (But since it's only for the initial sync it probably doesn't matter)

andresilva replied:
Yeah, I updated to change the order on this. 👍

expenses commented on bin/node/cli/src/service.rs:
```diff
@@ -99,6 +99,7 @@ macro_rules! new_full_start {
 					Some(Box::new(justification_import)),
 					None,
 					client,
+					select_chain,
```
@andresilva Do light clients _need_ a select chain instance? Is this something we could turn into an `Option`?

andresilva replied:
I think this can be turned into an option. This was added to fetch some data required for submitting equivocation reports, which the light clients won't do for now as only block authors submit them.
